### PR TITLE
Cluster

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-lash/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/reference/image_embedder.rst [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+------
+
+*********
+Inference
+*********
+
+The :class:`~flash.vision.ImageEmbedder` is already pre-trained on [ImageNet](http://www.image-net.org/), a dataset of over 14 million images.
+
+Use the :class:`~flash.text.ImageEmbedder` pretrained model for inference on any image tensor or image path using :func:`~flash.text.ImageEmbedder.predict`:
```
Comment (Borda) [inline docs/source/reference/image_embedder.rst:22]: `text`?
```suggestion
Use the :class:`~flash.vision.ImageEmbedder` pretrained model for inference on any image tensor or image path using :func:`~flash.vision.ImageEmbedder.predict`:
```

File: docs/source/reference/image_embedder.rst [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	print(predictions)
+
+Or on an image tensor
+
+.. code-block:: python
+
+	from flash.vision import ImageEmbedder
+
+	# Load finetuned task
+	embedder = ImageEmbedder(backbone='swav-imagenet')
```
Comment (Borda) [inline docs/source/reference/image_embedder.rst:42]:
```suggestion
Or on a random image tensor

.. code-block:: python
```

File: docs/source/reference/image_embedder.rst [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+*********
+Inference
+*********
+
+The :class:`~flash.vision.ImageEmbedder` is already pre-trained on [ImageNet](http://www.image-net.org/), a dataset of over 14 million images.
+
+Use the :class:`~flash.text.ImageEmbedder` pretrained model for inference on any image tensor or image path using :func:`~flash.text.ImageEmbedder.predict`:
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/reference/image_embedder.rst:24]: could it run on CPU? I would consider turning it to `.. testcode::` so we are on the safe side that it is working

File: docs/source/reference/image_embedder.rst [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    import flash
+    from flash import download_data
+    from flash.vision import ImageClassificationData, ImageClassifier
+
+    # 1. Download the data
+    download_data("https://pl-flash-data.s3.amazonaws.com/hymenoptera_data.zip", 'data/')
+
+    # 2. Load the data
+    datamodule = ImageClassificationData.from_folders(
+    	backbone="resnet18",
```
Comment (Borda) [inline docs/source/reference/image_embedder.rst:69]: why do I need to spec backbone here?

File: docs/source/reference/image_embedder.rst [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    trainer.save_checkpoint("image_classification_model.pt")
+
+------
+
+*********************
+Changing the backbone
+*********************
+By default, we use the encoder from `Swav <https://arxiv.org/pdf/2006.09882.pdf>`_ pretrained on Imagenet via contrastive learning. You can change the model run by the task by passing in a different backbone.
+
+.. note:: When changing the backbone, make sure you pass in the same backbone to the Task and the Data object!
```
Comment (Borda) [inline docs/source/reference/image_embedder.rst:97]:
```suggestion
.. note:: 

   When changing the backbone, make sure you pass in the same backbone to the Task and the Data object!
```

File: flash/core/model.py [hunk 6]
```diff
@@ -115,10 +115,12 @@ def predict(
             The post-processed model predictions
 
         """
-        data_pipeline = data_pipeline or self.data_pipeline
-        batch = x if skip_collate_fn else data_pipeline.collate_fn(x)
-        batch_x, batch_y = batch if len(batch) == 2 else (batch, None)
-        predictions = self.forward(batch_x)
+        self.eval()
+        with torch.no_grad():
```
Comment (Borda) [inline flash/core/model.py:119]: maybe you can set it as methods decorator...

File: flash/model_map.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    "SwAV": 'https://pl-bolts-weights.s3.us-east-2.amazonaws.com/swav/swav_imagenet/swav_imagenet.pth.tar',
+}
+
+
+def load_model(name):
+    if name == 'simclr-imagenet':
+        return load_simclr_imagenet()
+
+    elif name == 'swav-imagenet':
+        return load_swav_imagenet()
```
Comment (Borda) [inline flash/model_map.py:18]:
```suggestion
    if name in models:
        return models[name]()
```

File: flash_examples/predict/image_embedder.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import torch
+
+from flash.core.data import download_data
+from flash.vision import ImageEmbedder
+
+if __name__ == "__main__":
+
```
Comment (Borda) [inline flash_examples/predict/image_embedder.py:7]:
```suggestion
def main():
```

File: flash_examples/predict/image_embedder.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    assert embeddings.shape == torch.Size((1, 128))
+
+    # 5. Create a tensor random image
+    random_image = torch.randn(1, 3, 32, 32)
+
+    # 6. Generate an embedding from this random image
+    embeddings = embedder.predict(random_image)
+
+    # 7. Assert dimension
+    assert embeddings.shape == torch.Size((1, 128))
```
Comment (Borda) [inline flash_examples/predict/image_embedder.py:28]:
```suggestion
    assert embeddings.shape == torch.Size((1, 128))

if __name__ == "__main__":
    main()
```

File: docs/source/reference/image_embedder.rst (hunk 1 above)
Comment (tchaton) [reply docs/source/reference/image_embedder.rst:22]: Great catch !

File: docs/source/reference/image_embedder.rst (hunk 4 above)
Comment (tchaton) [reply docs/source/reference/image_embedder.rst:64]: Good catch !

File: flash/core/model.py (hunk 6 above)
Comment (tchaton) [reply flash/core/model.py:119]: Done !

File: flash_examples/predict/image_embedder.py (hunk 8 above)
Comment (tchaton) [reply flash_examples/predict/image_embedder.py:7]: I prefer to keep ` __name__ == "__main__"`

File: docs/source/reference/image_embedder.rst (hunk 1 above)
Comment (Borda) [reply docs/source/reference/image_embedder.rst]: but I think that this shall fail on docs build or test...

File: flash/core/model.py [hunk 10]
```diff
@@ -91,6 +92,17 @@ def test_step(self, batch: Any, batch_idx: int) -> None:
...
+    @contextmanager
+    def predict_context(self):
+        try:
+            self.eval()
+            torch.set_grad_enabled(False)
+            yield
+        finally:
+            self.train()
+            torch.set_grad_enabled(True)
+
```
Comment (carmocca) [inline flash/core/model.py:105]: This should be fixed in Lightning, right?

If so, please add a TODO

File: flash/model_map.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    model_config = {'model': swav.model, 'num_features': 3000}
+    return model_config
+
+
+_models = {'simclr-imagenet': load_simclr_imagenet, 'swav-imagenet': load_swav_imagenet}
+
+
+def load_model(name):
+    if not _BOLTS_AVAILABLE:
+        raise MisconfigurationException("Bolts isn't installed. Please, use pip install `lightning-bolts`.")
```
Comment (carmocca) [inline flash/model_map.py:29]:
```suggestion
        raise MisconfigurationException("Bolts isn't installed. Please, use `pip install lightning-bolts`.")
```

File: flash/model_map.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    if name in _models:
+        return _models[name]()
+    else:
+        raise MisconfigurationException("Currently, only `simclr-imagenet` and `swav-imagenet` are supported.")
```
Comment (carmocca) [inline flash/model_map.py:33]: Unnecessary else

```suggestion
    raise MisconfigurationException("Currently, only `simclr-imagenet` and `swav-imagenet` are supported.")
```

File: flash_examples/predict/image_embedder.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+    # 2. Create an ImageEmbedder with swav trained on imagenet.
+    # Check out SWAV: https://pytorch-lightning-bolts.readthedocs.io/en/latest/self_supervised_models.html#swav
+    embedder = ImageEmbedder(backbone="swav-imagenet", embedding_dim=128)
+
+    # 3. Generate an embedding from an image path.
+    embeddings = embedder.predict('data/hymenoptera_data/predict/153783656_85f9c3ac70.jpg')
+
+    # 4. Assert dimension
+    assert embeddings.shape == torch.Size((1, 128))
```
Comment (carmocca) [inline flash_examples/predict/image_embedder.py:19]: What is the purpose of these asserts? For the end user

File: flash/vision/embedding/image_embedder_model.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        image = torch.rand(32, 3, 32, 32)
+        embeddings = embedder(image)
+
+    """
+
+    def __init__(
+        self,
+        embedding_dim=None,
+        backbone="swav-imagenet",
+        pretrained=True,
```
Comment (carmocca) [inline flash/vision/embedding/image_embedder_model.py:92]: Add typing for these too

File: flash/model_map.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+def load_swav_imagenet(path_or_url: str = f"{ROOT_S3_BUCKET}/swav/swav_imagenet/swav_imagenet.pth.tar"):
+    swav = SwAV.load_from_checkpoint(path_or_url, strict=True)
+    model_config = {'model': swav.model, 'num_features': 3000}
+    return model_config
+
+
+_models = {'simclr-imagenet': load_simclr_imagenet, 'swav-imagenet': load_swav_imagenet}
+
+
+def load_model(name):
```
Comment (carmocca) [inline flash/model_map.py:27]: where is this used?

**Review by carmocca — COMMENTED** [review]
Can we have some tests? Also add the example to `test_examples` if it's not too slow

File: flash_examples/predict/image_embedder.py (hunk 13 above)
Comment (tchaton) [reply flash_examples/predict/image_embedder.py:19]: It was to demonstrate the shape was properly generated. But not really useful.

File: docs/source/reference/image_embedder.rst [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	embeddings = model.predict('path/to/image.png')
+	print(predictions)
+
+Or on a random image tensor
+
+.. testcode:: python
+
+	# 2. Perform inference on an image file
+	images = torch.rand(32, 3, 224, 224)
+	embeddings = model.predict(images)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply docs/source/reference/image_embedder.rst:41]: No, because the logic is DataPipeline.

File: flash/core/model.py [hunk 17]
```diff
@@ -91,6 +92,17 @@ def test_step(self, batch: Any, batch_idx: int) -> None:
...
+    @property
+    @contextmanager
+    def predict_context(self):
+        try:
+            self.eval()
+            torch.set_grad_enabled(False)
+            yield
+        finally:
+            self.train()
+            torch.set_grad_enabled(True)
```
Comment (carmocca) [inline flash/core/model.py:104]: why the `try...finally`?

```suggestion
        self.eval()
        torch.set_grad_enabled(False)
        yield
        self.train()
        torch.set_grad_enabled(True)
```

File: flash/vision/embedding/image_embedder_model.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+            self.head = nn.Sequential(
+                nn.Flatten(),
+                nn.Linear(num_features, embedding_dim),
+            )
+            rank_zero_warn('embedding_dim is not None. Remember to finetune first!')
+
+    def apply_pool(self, x):
+        if self.pooling_fn == torch.max:
+            x = self.pooling_fn(x, dim=-1)[0]
+            x = self.pooling_fn(x, dim=-1)[0]
```
Comment (carmocca) [inline flash/vision/embedding/image_embedder_model.py:139]: question: why [0] for max?

File: flash/core/model.py (hunk 17 above)
Comment (tchaton) [reply flash/core/model.py]: Moved to decorator.

File: flash/vision/embedding/image_embedder_model.py (hunk 18 above)
Comment (tchaton) [reply flash/vision/embedding/image_embedder_model.py:139]: max return argmax too.

**Review by carmocca — APPROVED** [review]

File: docs/source/reference/image_embedder.rst [hunk 19]
```diff
@@ -36,6 +36,8 @@ Or on a random image tensor
 
 .. testcode:: python
 
+  import torch
```
Comment (carmocca) [inline docs/source/reference/image_embedder.rst:40]: fix indentation

**Pull request merged**